### PR TITLE
build: check for backtrace(3) via FindBacktrace

### PR DESCRIPTION
**Pull request by dvrogozh**
Fixes: #1533

Change-Id: I81bcdb5e65c26f94b174af9b5e6218c14601c50f
Signed-off-by: Dmitry Rogozhkin <dmitry.v.rogozhkin@intel.com>

### CODE REVIEW

**Comment by bkuhls** [comment]
Building with a musl toolchain and libexecinfo works:

```
-- Looking for backtrace
-- Looking for backtrace - not found
-- Found Backtrace: /home/buildroot/buildroot/output/per-package/intel-mediadriver/host/x86_64-buildroot-linux-musl/sysroot/usr/lib/libexecinfo.so  
```

```
$ readelf -a output/per-package/intel-mediadriver/target/usr/lib/va/iHD_drv_video.so  | grep NEEDED
 0x0000000000000001 (NEEDED)             Shared library: [libexecinfo.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libigdgmm.so.12]
 0x0000000000000001 (NEEDED)             Shared library: [libstdc++.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libgcc_s.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libc.so]
```

**Comment by bkuhls** [comment]
Building with a glibc toolchain still works:
```
-- Looking for backtrace
-- Looking for backtrace - found
-- backtrace facility detected in default set of libraries
-- Found Backtrace: /home/buildroot/buildroot/output/per-package/intel-mediadriver/host/x86_64-buildroot-linux-gnu/sysroot/usr/include  
```
```
$ readelf -a output/per-package/intel-mediadriver/target/usr/lib/va/iHD_drv_video.so  | grep NEEDED
 0x0000000000000001 (NEEDED)             Shared library: [libigdgmm.so.12]
 0x0000000000000001 (NEEDED)             Shared library: [libstdc++.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libm.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libgcc_s.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libc.so.6]
```

**Review by XinfengZhang — APPROVED** [review]

**Comment by bkuhls** [comment]
Please do not merge this PR as is because building without libexecinfo on musl/uClibc is still broken due to mos_compat.h not being created, see my pending review.

**Comment by dvrogozh** [comment]
> building without libexecinfo on musl/uClibc is still broken due to mos_compat.h not being created, see my pending review.

Which comment? Above I see only comments that all works. Did you forget to post?

**Review by bkuhls — CHANGES_REQUESTED** [review]

**Comment by bkuhls** [comment]
> Which comment? Above I see only comments that all works. Did you forget to post?

Sorry, while using the Github review feature for the first time I did not notice that the state "Pending" meant that I still had to submit the review so others can see it ;)

File: media_softlet/linux/common/os/osservice/media_srcs.cmake [hunk 1]
```diff
@@ -18,12 +18,17 @@
 # ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 # OTHER DEALINGS IN THE SOFTWARE.
 
+if(Backtrace_FOUND)
+    configure_file(${CMAKE_CURRENT_LIST_DIR}/mos_compat.h.in ${CMAKE_BINARY_DIR}/mos_compat.h)
```
[the remark this answers is not in the document]
Comment (dvrogozh) [reply media_softlet/linux/common/os/osservice/media_srcs.cmake:22]: Ah, yes. You caught an issue in the PR which I overlooked. Yes, if backtrace is not found at all, then Backtrace_FOUND will be OFF and mos_conmpat.h won't be generated. That's wrong, idea was that this file should always be generated regardless of backtrace found/not-found status. And this file rules out internally whether backtrace was found or not. I'll fix.

File: media_softlet/linux/common/os/osservice/mos_compat.h.in [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+* OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+* ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+* OTHER DEALINGS IN THE SOFTWARE.
+*/
+
+#ifndef __MOS_COMPAT_H__
+#define __MOS_COMPAT_H__
+
+// clib implementations (musl, uClibc) might provide backtrace(3)
+// in different header files. So, we need to detect which header to be used.
```
[the remark this answers is not in the document]
Comment (dvrogozh) [reply media_softlet/linux/common/os/osservice/mos_compat.h.in]: updated comment

File: media_softlet/linux/common/os/osservice/media_srcs.cmake (hunk 1 above)
[the remark this answers is not in the document]
Comment (dvrogozh) [reply media_softlet/linux/common/os/osservice/media_srcs.cmake]: Fixed. Can you try again?

[the remark this answers is not in the document]
Comment (bkuhls) [reply media_softlet/linux/common/os/osservice/media_srcs.cmake]: Building with uClibc (which lacks backtrace(3)) and without libexecinfo works now:
```
$ readelf -a output/per-package/intel-mediadriver/target/usr/lib/va/iHD_drv_video.so  | grep NEEDED
 0x0000000000000001 (NEEDED)             Shared library: [libigdgmm.so.12]
 0x0000000000000001 (NEEDED)             Shared library: [libstdc++.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libgcc_s.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libc.so.0]
```
From my pov this PR can now be merged, thanks!

File: media_softlet/linux/common/os/osservice/mos_compat.h.in (hunk 2 above)
[the remark this answers is not in the document]
Comment (bkuhls) [reply media_softlet/linux/common/os/osservice/mos_compat.h.in]: thanks

File: media_softlet/linux/common/os/osservice/media_srcs.cmake (hunk 1 above)
[the remark this answers is not in the document]
Comment (dvrogozh) [reply media_softlet/linux/common/os/osservice/media_srcs.cmake]: Thank you for trying out. We will proceed with the merging this in.

**Review by bkuhls — APPROVED** [review]

**Comment by bkuhls** [comment]
Build error on "ci / gcc-10-gen12-off" is unrelated
```
2022-11-02T16:27:07.2595421Z /home/runner/work/media-driver/media-driver/media/media_softlet/agnostic/Xe_M_plus/Xe_LPM_plus_base/shared/mediacopy/media_render_copy_xe_lpm_plus_base.cpp:45:10: fatal error: igvpkrn_xe_hpg.h: No such file or directory
```

**Comment by dvrogozh** [comment]
Rebased to avoid a conflict with some recently merged change.

**Pull request merged**